core/cloudflow-akka-util/src/main/scala/cloudflow/akkastream/util/scaladsl/windowing/SessionWindows.scala

```diff
@@ -0,0 +1,215 @@
+      private var lastTS: Long      = -1                          // last ts
+      private var lastlocalTS: Long = -1                          // last local ts
+
+      // COnvert timings to miliseconds
+      private val inactivityMS = inactivity.toMillis
+
+      this.setHandlers(inlet, outlet, this)
+
+      // Pre Start - start timer
+      override def preStart(): Unit =
+        scheduleWithFixedDelay(shape, 0.millisecond, inactivityInterval)
+
+      // Post stop - cleanup
+      override def postStop(): Unit = {
+        buffer.clear()
+        offsets.clear()
+      }
+
+      // New event
+      override def onPush(): Unit = {
+        // Get element and its time
+        val event = grab(inlet)
+        val time  = time_extractor(event._1)
+
+        // Check whether this is a new window
+        if ((lastTS > 0) && (time > (lastTS + inactivityMS))) {
+          submitReadyWindows()
+        }
+
+        // Add event
+        updateData[T](event._1, buffer, maxsize)
+        offsets += event._2
+        lastTS = time
+        lastlocalTS = System.currentTimeMillis()
+
+        // Ask for new event
+        if (!hasBeenPulled(inlet)) pull(inlet)
+      }
+
+      // new window request
+      override def onPull(): Unit =
+        if (!hasBeenPulled(inlet)) pull(inlet)
+
+      // New timer event
+      override protected def onTimer(timerKey: Any): Unit = {
+        // Timer event, see if we can submit some waitinfg events
+        val time = System.currentTimeMillis()
+        if ((lastTS > 0) && ((time - lastlocalTS) > inactivityMS))
+          submitReadyWindows()
+      }
+
+      private def submitReadyWindows(): Unit = {
+        val commitable = CommittableOffsetBatch(offsets.clone())
+        println(s"Send committable $commitable")
+        val result = (buffer.clone(), commitable)
+        buffer.clear()
+        offsets.clear()
+        lastTS = -1
+        emit(outlet, result)
+      }
+    }
+}
+
+/**
+ * Session Window based on the session value
+ *
+ * This type of window treats determines session based on the session value
+ *
+ * @param session_extractor a function for retrieving event-time (in ms). Default is using processing time
+ * @param inactivity time of inactivity constituting end of the session
+ * @param maxsize max size for collected data. If the window length is greater then maxsize, only the last maxsize elements will be returned
+ * @param inactivityInterval time interval for inactivity testing.
+ *
+ * Starting time for the first window is calculated based on the time of the first recieved event.
+ */
+case class SessionValueWindow[T](session_extractor: T ⇒ String,
+                                 inactivity: FiniteDuration,
+                                 maxsize: Int = 1000,
+                                 inactivityInterval: FiniteDuration = 100.millisecond)
+    extends GraphStage[FlowShape[(T, Committable), (Seq[T], Committable)]] {
```

Comment:
Is it possible to make this generic and not use Committable from alpakka kafka?

Reply:
Well, this is what I was suggested to you a couple of month ago - to make commitable generic. Right now it is in alpakka.kafka. The moment you guys introduce generic Committable, will be happy to do is. At the moment, its the only offset representation, that is there